### ISSUE TITLE
Semantic Coloring

### ISSUE THREAD

**Issue by svenefftinge** [theia-ide/yang-lsp#138]
We are adding semantic coloring support to Theia, Xtext and the LSP in general.
@andreasjakobik and @huyuwen do you have any proposals for special syntax coloring you would like to see in Yang? 
It can use the entire AST and linking information.

**Comment by huyuwen** [comment]
Hi, 
According to RFC7950, YANG is a data modeling language designed to model configuration and state data manipulated by different protocols.
So there are a few important conceptions that are worth to be highlighted from Semantic perspective.
1.	Data Definition Node, Schema Node and Reusable Node
Data Def Node (leaf, leaf-list, list, container, anydata, anyxml) are subset of Schema Node, they really exists in the real final data tree.
In Schema node (choice, case and data def node), choice and case defines a set of alternatives of data, so not all cases will appear in the final data tree.
Reusable Node, (grouping: reusable def, uses: reusable reference), grouping only defines reusable data, but until it (grouping) is be ‘uses’ somewhere else, they (data def in grouping) will not appear in final data tree.
**_In short, when presenting data in yang modeling language, there has different kinds of data nodes to be highlighted, (normal data, alternative data, and reusable data)_**.
2.	Extendible module statement
Augment, Refine and Deviation statements have the ability to extend/impact (Add, Replace, Remove, Disable) existing modules. 
All of them can change specific existing module,  and then a new derived module will be used in runtime.
The difference between them is that they have their own extending target and extending ways.
**_Therefore, it is important to highlight these extending part of a module._**
3.	Conditional module statement
In yang modeling language, ‘when’, ‘if-feature’ could be used to define conditional schema node. ‘must’ are used to define constraint of some schema node and operation.
**_They all indicate that those nodes which applied with them (when, if-feature, must) are conditionally enabled/valid/available. 
So they are also valuable to be highlighted._**
4.	Interface statement
There are Action/RPC/Notification statement in yang modeling language. Apparently, they are quite different comparing with data nodes.
**_They are interface (Operation/Notification) definition of some models. It makes sense to distinguish them apart from data nodes._**
5.	Referenceable statement
Identify/Feature/Extension/TypeDef are other conceptional definition statement. They will only be valuable or meaningful when they are referenced from other schema node or interface/operation node.
**_It would be good that these kind of referenceable statement could be highlighted separately._**
6.	Other statement
Of cause, there are lot’s of other statements. From our experience, currently, ‘key’, ‘default’ could be considered to be highlighted. Meanwhile 'description' providers most important info for all kinds of nodes, so it is important to make 'description' much readable.

**Comment by svenefftinge** [comment]
Thank you @huyuwen!

**Comment by kittaakos** [comment]
I start working on the implementation. The upstream PR for the Xtext-based LS is here: https://github.com/eclipse/xtext-core/pull/815

**Comment by kittaakos** [comment]
Depends on https://github.com/theia-ide/yang-lsp/issues/142.

**Comment by kittaakos** [comment]
@huyuwen, I have a few questions regarding the requirements. Could you please help me with them? Thank you!

> In Schema node (choice, case and data def node),

What is a _data def node_?

> Reusable Node, (grouping: reusable def, uses: reusable reference), grouping only defines reusable data, but until it (grouping) is be ‘uses’ somewhere else, they (data def in grouping) will not appear in final data tree.

Does that mean we do not want to highlight them **iff** they're not being used from somewhere else?

> In yang modeling language, ‘when’, ‘if-feature’ could be used to define conditional schema node. ‘must’ are used to define constraint of some schema node and operation.
> They all indicate that those nodes which applied with them (when, if-feature, must) are conditionally enabled/valid/available.

Can you please clarify on this requirement? What exactly should be highlighted for the `when`, `if-feature`, and `must`? Assuming we have the following example with the `if-feature`, do we want always to highlight `"foo"` and `"bar"` from the `"foo and not bar"` or only **iff** some conditions are met? Let me know if I have misunderstood something.

`u.yang`:
```
module u {
  yang-version 1.1;
  namespace urn:u;
  prefix u;

  feature foo;
  feature bar;

  container a {
    if-feature "foo and not bar";
  }

  container t {
    notification x;
    action reset {
      input {
        leaf time {
          type string;
        }
      }
    }
  }

}
```

**Comment by andreasjakobik** [comment]
Hello, semantic coloring is a cool feature and I'm eager to see the results of this effort! That said, I suspect too many colors for different semantic meanings may risk being a little confusing to the user. Not sure usability or aesthetics benefit by default. Are there any studies in this area to learn from?
Anyway, looking forward to see the results, and we can judge ourselves. I assume the color used for each semantic part will be configurable?
Thanks a lot!
/Andreas

**Comment by kittaakos** [comment]
> I assume the color used for each semantic part will be configurable?

No, it will not be configurable, at least not within this task.

**Comment by huyuwen** [comment]
@kittaakos, here is my comments.
1. 'data def node' is just data definition node what I described above. So in short data definition node must be schema node, but schema node might not be data definition node, like choice, case.

2. Regarding to grouping node, I think, if a grouping is not 'uses' by other node, we don't need to apply semantic highlight for it. This is also an indicate that the grouping is an unused definition. In such case, yang module designer should consider to keep the grouping or remove it.

3. For 'if-feature', you are right. It needs additional input during yang module handling. As to your example, we can only conclude that container 'a' will be valid in some cases (depends on other input, like **_pyang --feature_** options), but not be always exist in yang data node tree.
   For 'when', it is like 'if-feature', but just considering information from other aspect (for example, during system runtime, when a system trying to load yang module).
   So simply to say, when a schema node has when/if-feature, the schema node will not always be valid/exists, it needs additional information to decide. I think that's very useful information for yang module designer and user to know. Agree?
   For 'must', it is one kind of constraint for schema node. If the defined constraint is violated, there should have some warning/error message. So it will be good to highlight a node when it has 'must'. It will reminder yang module designer or user, there might be waring/error message trigger by that node.
     Of cause, there is no need to analysis/evaluate these condition/constraint in detail, but just show those nodes which has conditional defintion and highlight them to yang designer and user.
